### DISCUSSION TITLE
Loki: Determine what show context query should be

### DISCUSSION THREAD

**Discussion by tomwilkie** [grafana/grafana#45685] [Observability]
When I search for an error code, and then click "show context", I don't see the preceeding stacktrace / log lines.  One reason I think this is happening is we're for entries in the same stream, and its quite common for a single target to output multiple streams now, eg broken down by log level.  We should consider only including the job and instance labels when searching for context, or alternatively excluding the log level label.  WDYT?

**Comment by aocenas** [comment]
@tomwilkie the context should show all the lines without the filter part. So `{label="val"} |="search"` is the query then context is `{label="val"}`. So this seems like the former query won't show the lines you want either is that right? Do you mean we should alter the query for the context a bit more? Would that be generic enough to do every time or more as a datasource config thing?

**Comment by tomwilkie** [comment]
> Do you mean we should alter the query for the context a bit more?

Yeah, we should do this for sure.  The use case is do a query `{foo="bar"} |= "baz"` to find something and then want to see what came before that thing in the stream, which would need you to remove the filer.

> Would that be generic enough to do every time or more as a datasource config thing?

I think generic enough to do it each time.  We should define context as being on a stream, not a query.

**Comment by aocenas** [comment]
> Yeah, we should do this for sure. The use case is do a query {foo="bar"} |= "baz" to find something and then want to see what came before that thing in the stream, which would need you to remove the filer.

Yeah that' what we should be doing right now.

1. If your query is `{foo="bar"} |= "baz"`, in context you should see all the logs for `{foo="bar"}`.

2. From the issue I thought you mean we should do a something like changing `{foo="bar", app="shmoo"} |= "baz"` to  `{foo="bar"}` in the context to include more logs.

So if 1. isn't working that would be a bug for 2. that would be a feature request and would probably need more thought on how to implement that generically if possible.

**Comment by tomwilkie** [comment]
> Yeah that' what we should be doing right now.

Oh okay, looks like we are.

Problem is I see this behaviour:

![Screenshot 2020-06-24 at 13 59 32](https://user-images.githubusercontent.com/444037/85560874-fee84680-b622-11ea-9300-86ef39cb8d21.png)

But I guess this is because we are sending the level label in the context query.   We really only want to send "target" labels, not labels extracted by promtail.  But Grafana doesn't know the difference.  @slim-bean any ideas here?

**Comment by slim-bean** [comment]
What if "show context" sent the query with all the labels defining the stream for the log line selected and removed any filter expressions.

**Comment by aocenas** [comment]
@slim-bean That's what we do right now and as @tomwilkie  says that is the problem because level="error" is part of that log line labels and we send that to loki in the context query. That's why it looks like we are sending also the filter but we don't.

Tbh I always thought that the context should be doing the same query user writes into the query input, just stripping the filter.

I get that showing narrower context of the particular line, so that you can see only lines logged by the same process for example would make sense but as @tomwilkie mentions we do not know which set of labels identify that and which should be stripped.
cc @davkal

**Comment by halochou** [comment]
Any update or workaround on this issue? I can still see this bug in Granfana 8.0

**Comment by christineywang** [comment]
+1 on an update. I have a user who can recreate this anomalous behavior.

**Comment by zackman0010** [comment]
I would also like to +1 the ability to filter out which labels the Show Context button uses. I'm currently seeing the same issue with the log level label hiding important context.
Specifically, when we search to find an error, the error itself is printed at ERROR level. The surrounding context for the error is printed at various log levels, but usually INFO and DEBUG. However, pressing Show Context on the error message shows only other ERROR level messages and not the actual context we need for debugging purposes.

**Reply by abh** [reply]
My use case is that I find a request from a stream labeled `stream: stdout` and then would like to get the context with all the same labels, except don't filter on the `stream` label to include `stderr` as well.

The only workaround I have figured out is to do the search (as I do now) and then manually construct the second query around the timestamp I found in the first search. Very cumbersome...

**Reply by rad-pat** [reply]
I have very similar issue as this. Trying to move to Loki logging from Kibana/Elasticsearch. The show context (View surrounding documents) they implement is exactly what I needed in the past and would be great to see similar implemented here.

**Comment by ivanahuckova** [comment]
The show context query is now editable, therefore I am closing this issue.
